### PR TITLE
tests: add test for "sql: adjust catalog state when updating items"

### PR DESCRIPTION
**Pull request by nrainer-materialize**
This adds a test for https://github.com/MaterializeInc/materialize/pull/21185.

DO NOT MERGE BEFORE https://github.com/MaterializeInc/materialize/pull/21185.

### CODE REVIEW

File: misc/python/materialize/checks/drop_subsource.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+from materialize.checks.checks import Check
+from materialize.checks.common import KAFKA_SCHEMA_WITH_SINGLE_STRING_FIELD
+
+
+class DropSubsource(Check):
+    def initialize(self) -> Testdrive:
+        return Testdrive(
+            dedent(
+                """
+                > CREATE SECRET pgpass AS 'postgres'
```
Comment (philip-stoev) [inline misc/python/materialize/checks/drop_subsource.py:22]: This is likely going to conflict with the PgCdc Check. You need a dedicated secret/connection/etc, everything.

File: misc/python/materialize/checks/drop_subsource.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                  ADD SUBSOURCE t3;
+                """,
+            ]
+        ]
+
+    def validate(self) -> Testdrive:
+        return Testdrive(
+            dedent(
+                """
+            ! DROP SOURCE t3;
```
Comment (philip-stoev) [inline misc/python/materialize/checks/drop_subsource.py:85]: This test is effective, I just checked against main and it fails .

File: misc/python/materialize/checks/drop_subsource.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        return [
+            Testdrive(dedent(s))
+            for s in [
+                """
+                $ postgres-execute connection=postgres://postgres:postgres@postgres
+                CREATE TABLE t3 (f1 INTEGER);
+                ALTER TABLE t3 REPLICA IDENTITY FULL;
+                INSERT INTO t3 VALUES (1);
+                
+                > ALTER SOURCE mz_source
```
Comment (philip-stoev) [inline misc/python/materialize/checks/drop_subsource.py:67]: Create two subsources here. You drop one in manipulate2 , and let the other persist until validate.

File: misc/python/materialize/checks/drop_subsource.py (hunk 2 above)
Comment (philip-stoev) [inline misc/python/materialize/checks/drop_subsource.py:85]: Here , you need two identical error checks -- one each for the subsources created in manipulate1 and in manipulate2

**Review by philip-stoev — APPROVED** [review]
Thank you. The test is effective in checking for this particular bug.

It does not cover the complete lifecycle of a subsource, with `INSERT` statements and `SELECTs` in validation, but it is appropriately named `DropSubsource` , and not `Subsource` in general.

File: misc/python/materialize/checks/drop_subsource.py
Comment (philip-stoev) [inline misc/python/materialize/checks/drop_subsource.py:1]: A more appropriate name for this file will be `subsource.py`, and then `DropSubsource` is just the first subsource-related Check we will have in it.

Comment (nrainer-materialize) [reply misc/python/materialize/checks/drop_subsource.py:1]: ✅

File: misc/python/materialize/checks/drop_subsource.py (hunk 2 above)
Comment (nrainer-materialize) [reply misc/python/materialize/checks/drop_subsource.py]: ✅

File: misc/python/materialize/checks/drop_subsource.py (hunk 3 above)
Comment (nrainer-materialize) [reply misc/python/materialize/checks/drop_subsource.py]: ✅ Added `t4`

File: misc/python/materialize/checks/drop_subsource.py (hunk 1 above)
Comment (nrainer-materialize) [reply misc/python/materialize/checks/drop_subsource.py]: Adjusted some, migtht need to address further objects.

**Comment by sploiselle** [comment]
@nrainer-materialize tysm

**Pull request merged**